### PR TITLE
chore: refactor `conf` of `manager api`

### LINKED ISSUE #662: chore(manager-api): use yaml as configuration file format
original commit: https://github.com/apache/apisix-dashboard/pull/612/files#r515739896

- Why do you submit this issue?

- [x] Feature or performance improvement

### LINKED ISSUE #677: discuss: refactor `conf` of `manager api`
relate to  #620

- [ ] Do not write absolute paths in the code

- [ ] Unified format, use snake case for config

### PR DESCRIPTION
**Pull request by nic-chen**
close  #677
close #662

- [x] Do not write absolute paths in the code`
- [x] Unified format, use snake case for config
- [x] use YAML as configuration file format

### CODE REVIEW

**Comment by nic-chen** [comment]
@membphis 

I dealt with the two issues together, because it is too troublesome to debug if divided into two PRs, especially the docker part, the build time is too long

**Comment by membphis** [comment]
> @membphis
> 
> I dealt with the two issues together, because it is too troublesome to debug if divided into two PRs, especially the docker part, the build time is too long

ok, that is fine.

I think we need a better title for this PR. and please take a look at this CI, failed: https://github.com/apache/apisix-dashboard/pull/693/checks?check_run_id=1356022130#step:7:222

File: api/conf/conf.go [hunk 1]
```diff
@@ -79,12 +82,19 @@ type Config struct {
 }
 
 func init() {
-	flag.StringVar(&confDir, "c", "./conf/", "conf dir")
-	flag.Parse()
+	//go test
+	if strings.HasSuffix(os.Args[0], ".test") {
```
Comment (membphis) [inline api/conf/conf.go:86]: I do not think that is a good way

File: api/conf/conf.go [hunk 2]
```diff
@@ -79,12 +82,19 @@ type Config struct {
...
+		_, basePath, _, _ := runtime.Caller(0)
+		confDir = filepath.Dir(basePath) + "/"
+		fmt.Println("confDir:", confDir)
+	} else {
+		flag.StringVar(&confDir, "c", "./conf/", "conf dir")
```
Comment (membphis) [inline api/conf/conf.go:91]: `flag.StringVar(&confDir, "c", "./conf/", "conf dir")`
to
`flag.StringVar(&confDir, "c", "conf/config.yaml", "config file path")`

Comment (membphis) [reply api/conf/conf.go:91]: as we talked right now, we should use `flag.StringVar(&workDir, "p", ".", "current work dir")`.

support to specify the prefix path via ENV, `APISIX_API_WORKDIR` seems a good name for us.

we should support both of them.

```sh
# style 1
manager-api -p /home/xxx/apisix/manager-api

# style 2
APISIX_API_WORKDIR=/home/xxx/apisix/manager-api manager-api
```

Comment (juzhiyuan) [reply api/conf/conf.go:91]: I would prefer style 1 IMO

Comment (juzhiyuan) [reply api/conf/conf.go:91]: I would prefer style 2 IMO

Comment (nic-chen) [reply api/conf/conf.go]: it's better. and user don't need to config other dirs or file paths.

File: api/conf/conf.go (hunk 1 above)
Comment (nic-chen) [reply api/conf/conf.go]: fixed.

File: api/conf/conf.go (hunk 2 above)
Comment (juzhiyuan) [reply api/conf/conf.go]: OK!

File: Makefile [hunk 3]
```diff
@@ -49,5 +49,5 @@ golang-lint: ## Run the golangci-lint application (install if not found)
 ### api-test:         Run the tests of manager-api
 .PHONY: api-test
 api-test:
-	cd api/ && go test -v -race -cover -coverprofile=coverage.txt -covermode=atomic ./...
+	cd api/ && APISIX_API_WORKDIR=$$PWD go test -v -race -cover -coverprofile=coverage.txt -covermode=atomic ./...
```
Comment (membphis) [inline Makefile:52]: I am confused with `./...`, can you explain this? @nic-chen

File: api/Dockerfile [hunk 4]
```diff
@@ -17,45 +17,34 @@
 
 FROM golang:1.13.8 AS build-env
 
-WORKDIR /go/src/github.com/apisix/manager-api
+RUN mkdir -p /usr/local/apisix-dashboard/api
```
Comment (membphis) [inline api/Dockerfile:20]: one PR for one thing, I think we can fix this in new PR

File: api/Dockerfile [hunk 5]
```diff
@@ -17,45 +17,34 @@
...
+WORKDIR /usr/local/apisix-dashboard/api
+
 COPY . .
-RUN mkdir -p /go/manager-api \
-    && mkdir -p /go/manager-api/build-tools \
+
+RUN mkdir -p ../output/conf \
+    && cp ./conf/* ../output/conf
+
+RUN wget https://github.com/api7/dag-to-lua/archive/v1.1.tar.gz -O /tmp/v1.1.tar.gz \
```
Comment (membphis) [inline api/Dockerfile:29]: change it from `/tmp/v1.1.tar.gz` to `/tmp/dag-to-lua.tar.gz`

File: api/conf/conf.yaml [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+# limitations under the License.
+#
+# PLEASE DO NOT UPDATE THIS FILE!
+# If you want to set the specified configuration value, you can set the new
+# value in the conf/config.yaml file.
+#
+
+conf:
+  listen:
+    host: 127.0.0.1
```
Comment (membphis) [inline api/conf/conf.yaml:24]: do we support `*` here?

File: api/conf/conf.yaml [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    port: 8080
+  dag_lib_path: ''
```
Comment (membphis) [inline api/conf/conf.yaml:26]: `''` empty string? is it a valid path?

File: api/internal/handler/route/route.go [hunk 8]
```diff
@@ -175,7 +175,7 @@ func generateLuaCode(script map[string]interface{}) (string, error) {
 	}
 
 	cmd := exec.Command("sh", "-c",
-		"cd "+conf.DagLibPath+" && lua cli.lua "+
+		"cd "+conf.WorkDir+"/dag-to-lua && lua cli.lua "+
```
Comment (membphis) [inline api/internal/handler/route/route.go:178]: seems wrong.

we should use `conf.dag_lib_path`, all right?

**Review by membphis — CHANGES_REQUESTED** [review]

File: Makefile (hunk 3 above)
Comment (nic-chen) [reply Makefile:52]: that means discover and run any tests within current directly or their subdirectories.

File: api/Dockerfile (hunk 4 above)
Comment (nic-chen) [reply api/Dockerfile:20]: ok,let's revert this file

File: Makefile (hunk 3 above)
Comment (membphis) [reply Makefile:52]: `./...` is it a file path?

File: api/conf/conf.yaml (hunk 7 above)
Comment (nic-chen) [reply api/conf/conf.yaml:26]: it should be removed.

File: api/internal/handler/route/route.go (hunk 8 above)
Comment (nic-chen) [reply api/internal/handler/route/route.go:178]: I'will remove that config.users should not need to this dir.

File: api/conf/conf.yaml (hunk 6 above)
Comment (nic-chen) [reply api/conf/conf.yaml:24]: don't support. but it support `localhost`. I name it from official comment

	// Addr optionally specifies the TCP address for the server to listen on,
	// in the form "host:port". If empty, ":http" (port 80) is used.
	// The service names are defined in RFC 6335 and assigned by IANA.
	// See net.Dial for details of the address format.

File: Makefile (hunk 3 above)
Comment (nic-chen) [reply Makefile:52]: no, not a file path

File: api/conf/conf.yaml [hunk 9]
```diff
@@ -14,12 +14,23 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
-version: '3'
+# PLEASE DO NOT UPDATE THIS FILE!
```
Comment (membphis) [inline api/conf/conf.yaml:17]: it makes me confused, please confirm we need those warning message

File: api/test/docker/manager-api-conf.yaml [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+# PLEASE DO NOT UPDATE THIS FILE!
+# If you want to set the specified configuration value, you can set the new
+# value in the conf/config.yaml file.
+#
+
+conf:
+  listen:
+    host: 0.0.0.0
+    port: 8080
+  dag_lib_path: ''
```
Comment (membphis) [inline api/test/docker/manager-api-conf.yaml:26]: do we need to delete this line too?

File: api/conf/conf.yaml [hunk 11]
```diff
@@ -14,12 +14,23 @@
...
-    command: go test -v github.com/apisix/manager-api/service
+conf:
+  listen:
+    host: 127.0.0.1
+    port: 8080
+  etcd:
+    endpoints:
+      - 127.0.0.1:2379
+authentication:
+  secret: secret
```
Comment (membphis) [inline api/conf/conf.yaml:30]: why we need this configuration?

File: api/conf/conf.yaml (hunk 9 above)
Comment (nic-chen) [reply api/conf/conf.yaml]: it is useless. i will remove them

File: api/test/docker/manager-api-conf.yaml (hunk 10 above)
Comment (nic-chen) [reply api/test/docker/manager-api-conf.yaml]: sure.

File: api/conf/conf.yaml (hunk 11 above)
Comment (nic-chen) [reply api/conf/conf.yaml:26]: it's for jwt token generation

Comment (membphis) [reply api/conf/conf.yaml:26]: need some comment here

Comment (nic-chen) [reply api/conf/conf.yaml:26]: all right

File: api/test/docker/manager-api-conf.yaml [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+  listen:
+    host: 0.0.0.0
+    port: 8080
+  etcd:
+    endpoints:
+      - 172.16.238.10:2379
+      - 172.16.238.11:2379
+      - 172.16.238.12:2379
+authentication:
+  secret: secret
```
Comment (membphis) [inline api/test/docker/manager-api-conf.yaml:28]: comments too

**Comment by membphis** [comment]
> api/test/docker/manager-api-conf.yaml

use the original `conf/config.yaml` + `sed` is a better way @nic-chen

**Comment by nic-chen** [comment]
> use the original `conf/config.yaml` + `sed`

It's not just used in CI. we run e2e test in local need it too. I think it may cost some time. we could fix it later.

File: api/test/docker/manager-api-conf.yaml (hunk 12 above)
Comment (nic-chen) [reply api/test/docker/manager-api-conf.yaml]: done.

File: api/test/docker/manager-api-conf.yaml [hunk 13]
```diff
@@ -15,20 +15,24 @@
...
 authentication:
-  secret: secret
-  expireTime: 3600
+  secret: secret             # secret for jwt token generation.
+                             # *NOTE*: Highly recommended to modify this value to protect `manager api`.
+                             # if it's default value, when `manager api` start , it will generate a random string to replace it.
+  expireTime: 3600           # jwt token expire time, in second
   users:
-    - username: admin
+    - username: admin        #username and password for login `manager api`
```
Comment (juzhiyuan) [inline api/test/docker/manager-api-conf.yaml:34]: space after #

**Review by juzhiyuan — CHANGES_REQUESTED** [review]

File: api/conf/conf.yaml [hunk 14]
```diff
@@ -17,16 +17,18 @@
...
 authentication:
-  secret: secret
-  expireTime: 3600
+  secret: secret             # secret for jwt token generation.
+                             # *NOTE*: Highly recommended to modify this value to protect `manager api`.
+                             # if it's default value, when `manager api` start , it will generate a random string to replace it.
+  expireTime: 3600           # jwt token expire time, in second
   users:
-    - username: admin
+    - username: admin        #username and password for login `manager api`
```
Comment (membphis) [inline api/conf/conf.yaml]: need a space, `# username ...`

File: api/conf/conf.yaml [hunk 15]
```diff
@@ -17,16 +17,18 @@
 
 conf:
   listen:
-    host: 127.0.0.1
-    port: 8080
+    host: 127.0.0.1          # `manager api` listening ip or host name
+    port: 8080               # `manager api` listening port
   etcd:
-    endpoints:
+    endpoints:               # it's possible to define multiple etcd hosts addresses of the same etcd cluster.
```
Comment (membphis) [inline api/conf/conf.yaml:23]: please use better comments, it is not easy to understand

File: api/test/docker/manager-api-conf.yaml [hunk 16]
```diff
@@ -15,20 +15,24 @@
 # limitations under the License.
 #
 
+
```
Comment (membphis) [inline api/test/docker/manager-api-conf.yaml:18]: I do not think that is a good way to copy the `api/conf/conf.yaml` again only for testing.

we should use `cp` + `sed` to generate `manager-api-conf.yaml`

**Review by membphis — CHANGES_REQUESTED** [review]

**Comment by membphis** [comment]
> It's not just used in CI. we run e2e test in local need it too. I think it may cost some time. we could fix it later.

ok, got it. ^_^

File: api/conf/conf.yaml (hunk 15 above)
Comment (nic-chen) [reply api/conf/conf.yaml:23]: I copy them from APISIX.

Comment (membphis) [reply api/conf/conf.yaml:23]: we need to update the APISIX too ^_^

Comment (membphis) [reply api/conf/conf.yaml:23]: `supports defining multiple etcd host addresses for an etcd cluster`

how about this?

**Comment by membphis** [comment]
@nic-chen please confirm we have fixed those points.

image

**Review by membphis — APPROVED** [review]
LGTM

**Review by gxthrj — APPROVED** [review]

File: api/conf/conf.yaml (hunk 15 above)
Comment (nic-chen) [reply api/conf/conf.yaml:23]: it's better.

**Review by juzhiyuan — APPROVED** [review]

**Pull request merged**